src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/pre_compute.cpp

```diff
@@ -0,0 +1,182 @@
+
+        return {Cos, Sin};
+    }
+}  // namespace
+
+ov::npuw::patterns::pre_compute::RopePatternLLama2::RopePatternLLama2() {
+    auto shape_of = opp::wrap_type<ov::op::v3::ShapeOf>({opp::any_input()});
+    auto gather = opp::wrap_type<ov::op::v8::Gather>({shape_of, opp::any_input(), opp::any_input()});
+    auto concat_1 = opp::wrap_type<ov::op::v0::Concat>({gather, opp::any_input(), opp::any_input()});
+    //here we can seen inverse frequencies as a parameter or constant depending on partitioner passes
+    auto inv_freq = opp::wrap_type<ov::op::v0::Constant>();
+    auto inv_freq_convert = opp::optional<ov::op::v0::Convert>({inv_freq->output(0)});
+    auto broadcast = opp::wrap_type<ov::op::v3::Broadcast>({inv_freq_convert, concat_1});
+    auto position_ids = opp::wrap_type<ov::op::v0::Parameter>();
+    auto unsqueeze = opp::wrap_type<ov::op::v0::Unsqueeze>({position_ids, opp::wrap_type<ov::op::v0::Constant>()});
+    auto convert = opp::wrap_type<ov::op::v0::Convert>({unsqueeze});
+    auto matmul = opp::wrap_type<ov::op::v0::MatMul>({broadcast, convert});
+    auto transpose = opp::wrap_type<ov::op::v1::Transpose>({matmul, opp::any_input()});
+    auto concat_2 = opp::wrap_type<ov::op::v0::Concat>({transpose, opp::any_input()});
+    auto sin_cos = opp::wrap_type<ov::op::v0::Sin, ov::op::v0::Cos>({concat_2});
+
+    pattern = sin_cos;
+    verifier_cb = [=](ov::pass::pattern::Matcher& m) {
+        auto& node_to_output = m.get_pattern_value_map();
+
+        this->matched_position_ids = node_to_output.at(position_ids).get_node_shared_ptr();
+
+        this->matched_concat =  node_to_output.at(concat_1).get_node_shared_ptr();
+        auto matched_concat_2 = node_to_output.at(concat_2).get_node_shared_ptr();
+
+        // TODO: maybe check for it's dims
+        this->matched_inv_freq = node_to_output.at(inv_freq).get_node_shared_ptr();
+
+        // locating sin and cos
+        auto concat_users = matched_concat_2->output(0).get_target_inputs();
+        if (concat_users.size() != 2) {
+            LOG_INFO("Rope sin_cos invalid at: "<< matched_concat_2->get_name());
+            return false;
+        }
+        auto concat_it = concat_users.begin();
+        auto cos_node = concat_it->get_node();
+        concat_it++;
+        auto sin_node  = concat_it->get_node();
+
+        if (ov::as_type<ov::op::v0::Sin>(sin_node) == nullptr) {
+            std::swap(sin_node, cos_node);
+        }
+        if (ov::as_type<ov::op::v0::Sin>(sin_node) == nullptr ||
+            ov::as_type<ov::op::v0::Cos>(cos_node) == nullptr) {
+            LOG_INFO("Rope sin_cos nodes are not valid: sin=" << sin_node->get_name() << ", cos=" << cos_node->get_name());
+            return false;
+        }
+        LOG_INFO("Rope sin_cos nodes found : sin=" << sin_node->get_name() << ", cos=" << cos_node->get_name());
+        this->matched_cos = cos_node->shared_from_this();
+        this->matched_sin = sin_node->shared_from_this();
```

Comment:
is this ok for a `verifier` to modify the object state? I believe it should have no side effects?

Reply:
switched to multimatcher - no need for storing sin_cos outpu nodes at all any more